post: npm query 📦 (#57)
cloudx-labs/posts: posts/navarroaxel/11.npm-query-for-installed-packages.md

---
title: Querying installed packages with npm query
published: true
description: Check your installed dependencies using CSS selectors format.
tags: 'javascript, webdev, node, react'
cover_image: ./assets/npm-query-cover.jpg
id: 1209907
---

> Do you know if you use npm dependencies with `postinstall` scripts? Do you have a package installed twice? Since version 8.18.0 we can use [`npm query`](https://docs.npmjs.com/cli/v8/commands/npm-query) to find packages in the `node_modules` directory matching a specific query.

Reviewer: ```suggestion
Do you know if you have any package that makes use of postinstall scripts? Do you have a package installed twice? Since version 8.18.0 we can use [`npm query`](https://docs.npmjs.com/cli/v8/commands/npm-query) to find packages in the `node_modules` directory matching a specific query.
```

This is helpful in identifying possible issues and fixing them.

## Why is a postinstall script important?

The `postinstall` script allows a package to run code when it's installed and this could be used in a malicious way. The npm security team receives [reports of malware](https://docs.npmjs.com/reporting-malware-in-an-npm-package) in the registry and works to keep the ecosystem safe, but did you check which dependencies are running in `postinstall` scripts? Do you trust them?

> To uninstall all dependencies with a `postinstall` script use the following command:

Reviewer 2: Shouldn't you show an example that lists all these dependencies and not just a command that uninstalls things? Or at least include some warning message. Somebody might copy and execute without noticing it's an uninstall script.

```bash
npm query ":attr(scripts, [postinstall])" | jq 'map(.name) | join("\n")' -r | xargs -I {} npm uninstall {}
```

💡 The `jq` program is essential to filter data in JSON format.

## Finding duplicate packages

Sometimes we find some errors on the DevTools' console when React or Angular is being loaded twice on our website. This happens when a dependency is incompatible with the [semver](https://semver.org) of the same package declared on our `package.json`. E.g. We declared `react@^18.2.0` in our `package.json` and some dependency has `react@^16.8.0 || ^17.0.0` as dependency in their own `package.json`.

We can find every installed copy of a specific package using:

```bash
npm query '#react'
```

Or, if we're looking for old versions of React

```bash
npm query '#react:semver(<18.0.0)'
```

🧠 Remember that the selectors for `npm query` follow the [CSS selectors format](https://www.w3schools.com/cssref/css_selectors.asp).

> ## Check the dependencies’ license

Reviewer 2: ```suggestion
## Check the dependency licenses
```
